### ISSUE TITLE
udp Stream Recording Stops at ~1.7GB

### ISSUE THREAD

**Issue by deanholtwork** [tsduck/tsduck#180]
Hello,

I am recording a number of MPTS delivered to my server via udp. All recordings stop when the file size reaches ~1.7GB. I have tried this on two different machines and the same happens.

I installed TSDuck on Centos 7 (64bits) using by downloading the rpm and installing.

The command I am using is: 

tsp -I ip -l 172.17.55.60 239.201.129.101:1234 -O file /streams1/ts1.ts

I have also turned on debug and get the following:

[root@localhost ~]# tsp -v --debug=4 -I ip -l 172.17.190.60 239.201.129.102:1234 -O file /streams1/ts1.ts
debug level set to 4
* tsp options:
  --add-input-stuffing: 0/0
  --bitrate: 0 b/s
  --bitrate-adjust-interval: 5,000 milliseconds
  --buffer-size-mb: 16,777,216 bytes
  --debug: 4
  --list-processors: 0
  --max-flushed-packets: 0
  --max-input-packets: 0
  --realtime: unknown
  --monitor: 0
  --verbose: 1
  Number of packet processors: 0
  Input plugin 1:
    Name: ip
    Type: input
    Arg[0]: "-l"
    Arg[1]: "172.17.190.60"
    Arg[2]: "239.201.129.102:1234"
  Output plugin 1:
    Name: file
    Type: output
    Arg[0]: "/streams1/ts1.ts"
* Debug[4]: debug level set to 4
trying to load /usr/bin/tsplugin_ip.so
symbol tspNewProcessor not found in /usr/bin/tsplugin_ip.so
trying to load /usr/bin/tsplugin_file.so
realtime: yes, using --max-input-packets 1,000 --max-flushed-packets 1,000
* Debug[4]: ip: debug level set to 4
* Debug[4]: file: debug level set to 4
* Debug: tsp: buffer size: 89,240 TS packets, 16,777,120 bytes
* Debug: ip: setting socket reuse address to 1
* Debug: ip: binding socket to 239.201.129.102:1234
* ip: joining multicast group 239.201.129.102:1234 from local address 172.17.190.60
* Debug[2]: ip: received UDP packet, source: 172.17.190.43:1234, destination: 239.201.129.102:1234
* Debug: ip: initial buffer load: 7 packets, 1,316 bytes
* ip: unknown input bitrate
* Debug: file: creating file /streams1/ts1.ts, fd=5, error_code=0
* Debug: ip: input thread started
* Debug: file: output thread started
* Debug[2]: ip: received UDP packet, source: 172.17.190.43:1234, destination: 239.201.129.102:1234
* Debug[2]: ip: received UDP packet, source: 172.17.190.43:1234, destination: 239.201.129.102:1234
* Debug[2]: ip: received UDP packet, source: 172.17.190.43:1234, destination: 239.201.129.102:1234
* Debug[2]: ip: received UDP packet, source: 172.17.190.43:1234, destination: 239.201.129.102:1234
* Debug[2]: ip: received UDP packet, source: 172.17.190.43:1234, destination: 239.201.129.102:1234
* Debug[2]: ip: received UDP packet, source: 172.17.190.43:1234, destination: 239.201.129.102:1234

I am capturing from an AppearTV IRD which is outputting the full MPTS from an incoming DVB-S/2 signal to multicast. In total I am capturing 13 transponders 239.201.129.101:1234 through 239.201.129.113:1234. One thing I have noticed using debug is that when the file size is ~1.7GB tsp stops capturing the input packets.

It doesn't matter if I try and capture 1 stream or all 13 I get the same results.

Do you have any ideas? The only thing I can find on the internet is about Large File Support. I have no idea if this is the issue.

Regards,
Dean

**Comment by lelegard** [comment]
Hi @deanholtwork,

The problem is unusual. I have never seen it.

TSDuck is used to manipulate huge transport stream files, above 10 GB sometimes. So, 1.7 GB is not a limit of anything.

Is the output file size exactly the same at all times ? Which size in bytes ? Or is it around 1.7 GB only ?

The Large File Support is explicitly enabled inside TSDuck. On Linux, I usually use Fedora (sometimes Ubuntu) for serious testing. I only run the test suite on CentOS. But I do not think that using CentOS makes a difference. Did you try with other distro or OS, Fedora, Ubuntu, macOS, Windows ?

Is there any specific message when tsp stops at 1.7 GB ? Could you try to capture the end of a debug trace ? To avoid storing GB of logs, try something like:

~~~
tsp --debug=2 --synchronous-log ...  2>&1 | tail -1000 >test.log
~~~

I have one side remark, most likely absolutely unrelated to this problem: When you multicast different streams on the same network, it is better to use distinct ports in addition to distinct multicast addresses. Functionally, it does not change anything. But, when several processes on the same machine are receiving different streams on the same port, they all receive all streams at the same time and then have to filter the right stream. If there is only one receiver on the machine, the load is 1. With 2 receivers, the load is 2x2. With 3 receivers, it is 3x3. And so on, exponentially. But, again, this is just an unrelated remark.

**Comment by deanholtwork** [comment]
Hi Lelegard,

Thanks for the prompt reply.

I have done some further investigating this morning.

I disabled SELINUX and turned off firewalld -> same problem still happens

Captured a different stream on the same system. This streams is a SPTS at about 6mbps. tsp captures the stream and stops at a files size of approx 218MB. This leads me to think it is something to do with how long tsp thinks it has to record for or some other time related factor. 

Does tsp have a default recording time that I am missing a parameter to override this in the command below?

tsp -I ip -l 172.17.190.60 239.201.129.102:1234 -O file /streams1/ts2.ts

Regards,
Dean

**Comment by lelegard** [comment]
There is no specific limit in size or time in tsp. More precisely, there are dedicated plugins for that purpose. But a straightforward command as yours will normally run forever (at least until the output disk is full).

It would be interesting to understand how tsp decides to stop. Could you try the test I suggested; run with full debug and save only the last 1000 lines of debug output. Please post the last lines (maybe not 1000, just the final part).

Could you also make the same test on another machine with another distro or OS? Just to check if the problem is specific to your machine.

**Comment by deanholtwork** [comment]
OK. I have tried this on 4 different machines using a mixture of Centos 7, Fedora 29 and Ubuntu 10.04LTS

Ubuntu:
# tsp --debug=2 --synchronous-log -l 172.17.190.65 239.201.129.105:1234 -O file /streams1/ts5.ts

* Debug[2]: ip: received UDP packet, source: 172.17.190.60:54184, destination: 239.201.129.105:1234
* Debug[2]: ip: received UDP packet, source: 172.17.190.60:54184, destination: 239.201.129.105:1234
* Debug[2]: ip: received UDP packet, source: 172.17.190.60:54184, destination: 239.201.129.105:1234
* Debug[2]: ip: received UDP packet, source: 172.17.190.60:54184, destination: 239.201.129.105:1234
* Debug: Report logging thread terminated

Centos 7:
# tsp --debug=2 --synchronous-log -l 172.17.190.41 239.201.129.105:1234 -O file /streams1/ts5.ts

* Debug[2]: ip: received UDP packet, source: 172.17.190.60:54184, destination: 239.201.129.105:1234
* Debug[2]: ip: received UDP packet, source: 172.17.190.60:54184, destination: 239.201.129.105:1234
* Debug[2]: ip: received UDP packet, source: 172.17.190.60:54184, destination: 239.201.129.105:1234
* Debug[32649]: *** Internal error, thread aborted: std::bad_alloc

I ran the command again on Centos 7 and this is the output

* Debug[2]: ip: received UDP packet, source: 172.17.190.60:54184, destination: 239.201.129.105:1234
* Debug[2]: ip: received UDP packet, source: 172.17.190.60:54184, destination: 239.201.129.105:1234
* Debug[2]: ip: received UDP packet, source: 172.17.190.60:54184, destination: 239.201.129.105:1234
* Debug: Report logging thread terminated

Both times when the logging was terminated the stream stopped recording.

I didn't capture the logs from Fedora before I overwrote that with Ubuntu

**Comment by lelegard** [comment]
Thanks for the new tests.

> Both times when the logging was terminated the stream stopped recording.

In fact, this is the opposite. The `tsp` process stops for some reason that we need to elucidate and then, because the application is stopping, the logging thread also terminates as part of the global application shutdown.

The only conclusion we can make so far is that the `tsp` process properly terminates and is not simply killed from outside.

There is one exception with this message:

> *** Internal error, thread aborted: std::bad_alloc

In this case, the application terminates because of a thread terminated with a memory allocation failure exception. Either there is a memory leak in some plugin or the system is short of memory. I suggest to use the option `tsp -m` (monitor). Messages are regularly displayed with the CPU and virtual memory usage of the process. We will see if the memory stabilises or always increases. In the latter case, there is a memory leak we need to investigate further.

This problem is reported only once out of three. In the two other cases, in the log file, can you locate the same message, somewhere before the end? In the case where we see the message at the end, it is quite probable that the thread which crashed was the logging thread and it is normal that no other message is displayed afterwards. But, if the process runs out of virtual memory, any thread may stop because of a memory allocation. In that case, you may find a similar message somewhere above. In all cases, if there is a memory leak, `tsp -m` will detect it.

**Comment by deanholtwork** [comment]
The references to Centos 7 in my last post were incorrect. It should have been Fedora 29. I have 4 machines running various OS with tsp installed. Sorry for any confusion.

I ran

tsp --monitor -I ip -l 172.17.190.<xxx> 239.201.129.105:1234 -O file /streams1/ts5.ts

Ubuntu 18.04TLS:
root@dvb1:~# free -m
              total        used        free      shared  buff/cache   available
Mem:          15916         206       15008           1         701       15416
Swap:          4095           0        4095
root@dvb1:~# tsp --monitor -I ip -l 172.17.190.65 239.201.129.105:1234 -O file /streams1/ts5.ts
* [MON] 2018/12/19 06:56, resource monitoring started
* [MON] 2018/12/19 06:56, VM:414 MB (stable), CPU:6.80% (average:6.80%)
* [MON] 2018/12/19 06:56, VM:414 MB (stable), CPU:7.10% (average:6.95%)
* [MON] 2018/12/19 06:56, VM:414 MB (stable), CPU:7.10% (average:7.00%)
* [MON] 2018/12/19 06:56, VM:414 MB (stable), CPU:7.00% (average:6.99%)
* [MON] 2018/12/19 06:57, VM:414 MB (stable), CPU:7.00% (average:6.99%)
* [MON] 2018/12/19 06:57, VM:414 MB (stable), CPU:7.40% (average:7.06%)
* [MON] 2018/12/19 06:57, VM:414 MB (stable), CPU:6.80% (average:7.02%)
* [MON] 2018/12/19 06:57, VM:414 MB (stable), CPU:7.40% (average:7.07%)
* [MON] 2018/12/19 06:57, VM:414 MB (stable), CPU:7.10% (average:7.07%)
* [MON] 2018/12/19 06:57, VM:414 MB (stable), CPU:7.00% (average:7.06%)
* [MON] 2018/12/19 06:58, VM:414 MB (stable), CPU:6.30% (average:6.99%)
* [MON] 2018/12/19 06:58, VM:414 MB (stable), CPU:7.30% (average:7.02%)
* [MON] 2018/12/19 06:59, VM:414 MB (stable), CPU:6.95% (average:6.99%)
* [MON] 2018/12/19 07:00, VM:414 MB (stable), CPU:6.73% (average:6.93%)
* [MON] 2018/12/19 07:01, VM:414 MB (stable), CPU:6.88% (average:6.92%)
* [MON] 2018/12/19 07:02, VM:414 MB (stable), CPU:4.96% (average:6.59%)
* [MON] 2018/12/19 07:03, VM:414 MB (stable), CPU:0.00% (average:5.65%)
* [MON] 2018/12/19 07:04, VM:414 MB (stable), CPU:0.00% (average:4.94%)
* [MON] 2018/12/19 07:05, VM:414 MB (stable), CPU:0.00% (average:4.39%)
* [MON] 2018/12/19 07:06, VM:414 MB (stable), CPU:0.00% (average:3.95%)
* [MON] 2018/12/19 07:08, VM:414 MB (stable), CPU:0.00% (average:3.29%)
* [MON] 2018/12/19 07:10, VM:414 MB (stable), CPU:0.00% (average:2.82%)

The stream stopped recording at 07:01

Fedora 29
[root@localhost ~]# free -m
              total        used        free      shared  buff/cache   available
Mem:           7928         332        7184           1         411        7346
Swap:          8059           0        8059
[root@localhost ~]# tsp --m -I ip -l 172.17.190.41 239.201.129.105:1234 -O file /streams1/ts5.ts
tsp: ambiguous option --m (--max-flushed-packets, --max-input-packets)
[root@localhost ~]# tsp --monitor -I ip -l 172.17.190.41 239.201.129.105:1234 -O file /streams1/ts5.ts
* [MON] 2018/12/19 01:42, resource monitoring started
* [MON] 2018/12/19 01:42, VM:319 MB (stable), CPU:7.39% (average:7.39%)
* [MON] 2018/12/19 01:42, VM:319 MB (stable), CPU:8.00% (average:7.69%)
* [MON] 2018/12/19 01:42, VM:319 MB (stable), CPU:8.00% (average:7.79%)
* [MON] 2018/12/19 01:42, VM:319 MB (stable), CPU:5.30% (average:7.17%)
* [MON] 2018/12/19 01:43, VM:319 MB (stable), CPU:5.80% (average:6.89%)
* [MON] 2018/12/19 01:43, VM:319 MB (stable), CPU:6.70% (average:6.86%)
* [MON] 2018/12/19 01:43, VM:319 MB (stable), CPU:7.00% (average:6.88%)
* [MON] 2018/12/19 01:43, VM:319 MB (stable), CPU:7.00% (average:6.89%)
* [MON] 2018/12/19 01:43, VM:319 MB (stable), CPU:7.00% (average:6.91%)
* [MON] 2018/12/19 01:43, VM:319 MB (stable), CPU:7.00% (average:6.91%)
* [MON] 2018/12/19 01:44, VM:319 MB (stable), CPU:6.90% (average:6.91%)
* [MON] 2018/12/19 01:44, VM:319 MB (stable), CPU:7.00% (average:6.92%)
* [MON] 2018/12/19 01:45, VM:319 MB (stable), CPU:6.98% (average:6.94%)
* [MON] 2018/12/19 01:46, VM:319 MB (stable), CPU:7.00% (average:6.95%)
* [MON] 2018/12/19 01:47, VM:319 MB (stable), CPU:7.05% (average:6.97%)
* [MON] 2018/12/19 01:48, VM:319 MB (stable), CPU:5.65% (average:6.75%)
* [MON] 2018/12/19 01:49, VM:319 MB (stable), CPU:0.00% (average:5.79%)
* [MON] 2018/12/19 01:50, VM:319 MB (stable), CPU:0.00% (average:5.06%)
* [MON] 2018/12/19 01:51, VM:319 MB (stable), CPU:0.00% (average:4.50%)
* [MON] 2018/12/19 01:52, VM:319 MB (stable), CPU:0.00% (average:4.05%)
* [MON] 2018/12/19 01:54, VM:319 MB (stable), CPU:0.00% (average:3.37%)
* [MON] 2018/12/19 01:56, VM:319 MB (stable), CPU:0.00% (average:2.89%)
* [MON] 2018/12/19 01:58, VM:319 MB (stable), CPU:0.00% (average:2.53%)
* [MON] 2018/12/19 02:00, VM:319 MB (stable), CPU:0.00% (average:2.25%)
* [MON] 2018/12/19 02:02, VM:319 MB (stable), CPU:0.00% (average:2.02%)

The stream stopped recording at 01:48

This may not be related to the issue at all but on both systems the monitor messages were displayed approximately every 15 seconds when tsp initially started. After a couple of minutes the interval between messages started to increase.

**Comment by lelegard** [comment]
So, the monitoring reports a very stable memory usage. There is no memory leak (I wasn't aware of any, but who knows).

I see no reason for the process to stop. Did you log the last 1000 lines of the debug output as I suggested? Could you attach the corresponding text file? (Not inline the 1000 lines in a post) Maybe add the option `--timed-log` so that each log and debug message has a time stamp.

> This  may not be related to the issue at all but on both systems the monitor  messages were displayed approximately every 15 seconds when tsp  initially started. After a couple of minutes the interval between  messages started to increase.

This is normal, because the need for monitoring info depends on the run time, very fast at the start of the application, then slower and slower:

- Up to start + 2 mn, log every 10 seconds
- Up to start + 10 mn, log every minute
- Up to start + 20 mn, log every 2 minutes
- Up to start + 1 hour, log every 5 minutes
- After start + 1 hour, log every 30 minutes

This is what we expect from a short command and a server, respectively.

**Comment by deanholtwork** [comment]
[ubuntu1804LTS_2000.log](https://github.com/tsduck/tsduck/files/2694860/ubuntu1804LTS_2000.log)
[fc29_2000.log](https://github.com/tsduck/tsduck/files/2694862/fc29_2000.log)

Attached are the log files for the Ubuntu and Fedora machines.

Interestingly after the errors occurred, on both machines, if I type:
# ps -A | grep tsp

tsp is listed as a process and I can kill it with:
# kill <process_number>

**Comment by lelegard** [comment]
I see nothing specific in these logs, only packets being received. On Fedora, a thread (probably the logging one) terminates with a memory allocation failure while the monitoring shows a stable memory. On Ubuntu, the process seems to terminate properly but without evidence of the reason.

I cannot explain why there is a memory allocation failure inside the logging thread with a stable virtual memory. I need to reproduce the issue to analyze it. But, in many configurations, `tsp` runs for days - even weeks - and remain stable. I have never seen `tsp` crashing just because it is running for too long. So, I have not much hope to reproduce this unusual situation with such a simple command.

**Comment by deanholtwork** [comment]
OK. Understood.

I am beginning to think it is something to do with the setup/configuration of Linux machines.

I have to add the following line to /etc/sysctl.conf on all my linux machines to allow the multicast in

net.ipv4.conf.<interface-name>.force_igmp_version = 2

If I don't have that line in /etc/sysctl.conf or set it to net.ipv4.conf.<interface-name>.force_igmp_version = 0

I cannot receive the multicast. I have a Windows PC on the same network running VLC and I can watch the multicast video.

**Comment by deanholtwork** [comment]
I have everything working now. As you know there is nothing wrong with your software. It was totally down to me not configuring the servers correctly.

I hadn't enabled the igmp protocol in firewalld

I also needed to add the following to sysctl.conf

net.ipv4.conf.default.rp_filter=0
net.ipv4.conf.all.rp_filter=0
net.ipv4.conf.eno4.rp_filter=0
net.ipv4.icmp_echo_ignore_broadcasts=0

I had added some of them but not all.

Needless to say I am a lot wiser now if server configuration for multicast.

Many thanks for your support and sorry to have wasted you time.

Regards,
Dean

**Comment by lelegard** [comment]
Hi @deanholtwork 

Thank you for the feedback. There is no waste of time when solutions are properly documented since they are likely to be used by other users with similar issues.

Just to make it clear, could you confirm that the configuration you describe (igmp in firewalld and updates in sysctl.conf) had to be applied on the _server_ sending the packets only and that no specific configuration was required on _clients_ running TSDuck ?

**Comment by deanholtwork** [comment]
Hi @lelegard 

The configuration is for a server that is capturing multicast transport streams. For example:

tsp -I ip -l <nic_ip_address> <multicast_ip_address:port> -O file <filename.ts>

I also used TSDuck to playout a transport stream. I didn't make any configuration changes to make it play out the multicast transport streams. I will make the same configuration changes to the playout server that are required to the recording server and check to make sure everything works properly. I will let you know the outcome.

**Comment by joolzg** [comment]
ubuntu 16.04
tsp 3.16-1098
cli: tsp -I ip --local-address 192.168.1.17 224.170.1.0:3245 -O file temp.ts

Stopped when the file hit 2GB

**Comment by lelegard** [comment]
I tried to reproduce the problem without luck.

I tried on Ubuntu 18.10. This is a VMware virtual machine with a standard installation. Its main purpose is to build TSDuck on Ubuntu and I only installed the build environment. Nothing special was configured or installed. Note that I no longer have any Ubuntu 16.x and I do not have physical machines running Ubuntu, only VM's.

The host is a Mac. I multicast at full speed on the private virtual network from the Mac host:
~~~
$ tsversion
3.16-1098
$ tsp -I null -O ip 224.170.1.0:3245 --local-address 192.168.233.1
~~~

On the Ubuntu VM, I use a command which is almost identical to yours (only the local address changes, of course):
~~~
$ tsversion
3.16-1098
$ tsp -I ip --local-address 192.168.233.11 224.170.1.0:3245 -O file temp.ts
~~~

The size of `temp.ts` increases very rapidly, well beyond 2GB. After 3.2 GB, I killed the command.

So, this behaviour is not reproducible at will. It probably depends on some local system configuration. It does not mean that there is nothing to do with TSDuck. Maybe there is a specific situation or configuration that TSDuck does not handle properly. But that situation remains to be identified. I cannot reproduce it. It can be analysed only on systems where the situation occurs.

**Comment by joolzg** [comment]
Have you checked that your sending and responding to IGMP group messages as the default timeout is 125 seconds, maybe your losing the feed due to this not being kept open.

**Comment by lelegard** [comment]
Since everything is done on a LAN (and a virtual one), there is no router between the sender and receiver. IGMP requests are not necessary. Could you try with the sender and receiver on the same LAN? Just to check if the problem occurs only with a router in between.

With the router being used, could you capture all IGMP messages using Wireshark for instance?

**Comment by joolzg** [comment]
What data rate was the stream when you got the 1.7GB file, see if this divides down to 290.

298MB / 6mbps = 290 seconds

Also try

tcpdump -v -i xxx igmp

**Comment by lelegard** [comment]
> Stopped when the file hit 2GB

Something isn't clear here.
- Does the `tsp` command _terminate_ when the file reaches 2 GB?
- Or does the `tsp` command still runs but the output file size does no longer increase?

In the former case, IGMP timers or multicast routing have nothing to do with the problem. No longer receiving packets never makes `tsp` terminate. It simply waits for more packets (possibly forever).

In the latter case, on the other hand, the `tsp` command simply no longer receive packets and we may suspect something about multicast routing.

**Comment by joolzg** [comment]
No i stopped it as it had passed the 1.7GB and was way over 2GB

**Comment by joolzg** [comment]
ok just hit 3GB and climbing

**Comment by lelegard** [comment]
Hi @deanholtwork 

Can you confirm you still have the problem?
Did you find an explanation?

**Comment by deanholtwork** [comment]
Hi @lelegard, @joolzg 

I have fixed the issues I was having. As I mentioned above it was a combination of things.

1. Add igmp as a protocol to firewalld

2. Add the following to sysctl.conf
    net.ipv4.conf.default.rp_filter=0
    net.ipv4.conf.all.rp_filter=0
    net.ipv4.conf.eno4.rp_filter=0
    net.ipv4.icmp_echo_ignore_broadcasts=0

Change eno4 to match the interface you are capturing traffic on.

I had to reboot the server after making the changes to sysctl.conf as sysctl -p didn't make all the changes.
